### PR TITLE
TypeScript for animations

### PR DESCRIPTION
**Pull request by piaskowyk**
## Description

Refactor and transfter of animations module to TypeScript.

### CODE REVIEW

**Review by Szymon20000 — APPROVED** [review]

File: src/reanimated2/animation/decay.ts [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+import {
+  Animation,
+  AnimationCallback,
+  PrimitiveValue,
+  Timestamp,
+} from './commonTypes';
+import { Platform } from 'react-native';
+
+interface DecayConfig {
+  [key: string]: any;
```
Comment (jakub-gonet) [inline src/reanimated2/animation/decay.ts:11]: Why we allow unknown properties here?

File: src/reanimated2/animation/decay.ts [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+        ) {
+          animation.current = config.clamp[1];
+          return true;
+        }
+      }
+
+      if (Math.abs(v) < VELOCITY_EPS) {
+        return true;
+      }
+      return false;
```
Comment (jakub-gonet) [inline src/reanimated2/animation/decay.ts:87]: `return Math.abs(v) < VELOCITY_EPS`

File: src/reanimated2/animation/decay.ts [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+          if (config.clamp.length !== 2) {
+            console.error(
+              `clamp array must contain 2 items but is given ${config.clamp.length}`
+            );
+          }
+        } else {
+          console.error(
+            `config.clamp must be an array but is ${typeof config.clamp}`
+          );
+        }
```
Comment (jakub-gonet) [inline src/reanimated2/animation/decay.ts:102]: Maybe not in scope of this PR but we could invert those ifs to flatten this a bit:
```js
if (config.clamp) {
	if (!Array.isArray(config.clamp)){
		throw //typeof error
	}
	if(config.clamp.length !== 2){
		throw //length error
	}
}
```

File: src/reanimated2/animation/decay.ts [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+    return {
+      onFrame: decay,
+      onStart,
+      callback,
+      velocity: config.velocity || 0,
+      initialVelocity: 0,
+      current: 0,
+      lastTimestamp: 0,
+      startTimestamp: 0,
+    } as DecayAnimation;
```
Comment (jakub-gonet) [inline src/reanimated2/animation/decay.ts:132]: Why we have to cast it? Return type of the function should be typed at the function signature.

File: src/reanimated2/animation/decay.ts [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+      animation.startTimestamp = now;
+      animation.initialVelocity = config.velocity;
+      validateConfig();
+    }
+
+    return {
+      onFrame: decay,
+      onStart,
+      callback,
+      velocity: config.velocity || 0,
```
Comment (jakub-gonet) [inline src/reanimated2/animation/decay.ts:127]:
```suggestion
      velocity: config.velocity ?? 0,
```

File: src/reanimated2/animation/delay.ts [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+      callback,
+      previousAnimation: null,
+      startTime: 0,
+      started: false,
+    } as DelayAnimation;
+  });
+}
+
+/* Deprecated section, kept for backward compatibility. Will be removed soon */
+export function delay(
```
Comment (jakub-gonet) [inline src/reanimated2/animation/delay.ts:88]:
```js
/**
 * @deprecated Kept for backward compatibility. Will be removed soon.
 */
```

Comment (jakub-gonet) [reply src/reanimated2/animation/delay.ts:88]: IDEs highlight JSDoc comments so we may use them.

File: src/reanimated2/animation/index.ts [hunk 7]
```diff
@@ -0,0 +1,9 @@
+export * from './commonTypes';
+export * from './util';
+export * from './timing';
+export * from './spring';
+export * from './decay';
+export * from './delay';
+export * from './repeat';
+export * from './sequence';
+export * from './styleAnimation';
```
Comment (jakub-gonet) [inline src/reanimated2/animation/index.ts:9]: I'm not the fan of this kind of imports. We started to export more types to share them between modules and those shouldn't be used by end users. IMO, the better approach would be to explicitly list things we want to export so we have more control of it.

File: src/reanimated2/animation/util.ts [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+    } else if (typeof value === 'string') {
+      prefNumberSuffOnStart(animation, value, timestamp, previousAnimation);
+      animation.onFrame = prefNumberSuffOnFrame;
+      return;
+    }
+    baseOnStart(animation, value, timestamp, previousAnimation);
+  };
+}
+
+type ConditionalyStarting<
```
Comment (jakub-gonet) [inline src/reanimated2/animation/util.ts:185]: I'm not fan of this name. Also, it has a typo.

File: src/reanimated2/commonTypes.ts [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+
+export interface StyleProps extends ViewStyle, TextStyle {
+  originX?: number;
+  originY?: number;
+  [key: string]: any;
+}
+
+export interface AnimatedStyle
+  extends Record<string, Animation<AnimationObject>> {
+  [key: string]: any;
```
Comment (jakub-gonet) [inline src/reanimated2/commonTypes.ts:43]: We have the type for AnimatedStyle IIRC.

File: src/reanimated2/layoutReanimation/index.ts [hunk 10]
```diff
@@ -1,3 +1,4 @@
-export * from './AnimatedRoot';
+require('./AnimatedRoot');
+export * from './AnimatedLayout';
```
Comment (jakub-gonet) [inline src/reanimated2/layoutReanimation/index.ts:2]: Why we do that here?

File: src/reanimated2/animation/util.ts [hunk 11]
```diff
@@ -182,7 +182,7 @@ export function decorateAnimation<
   };
 }
 
-type ConditionalyStarting<
+type AniamtionToDecoration<
```
Comment (jakub-gonet) [inline src/reanimated2/animation/util.ts:185]: You have a typo here.

**Pull request merged**